### ISSUE TITLE
Add object store level support for counters

### ISSUE THREAD

**Issue by austinzheng** [realm/realm-object-store#357]
c.f. https://github.com/realm/realm-sync/issues/16

It seems like making counters a type apart from `Int` is a good idea, as originally suggested in the linked ticket. Although the underlying engine supports a type which mixes 'set' and 'add' operations, it's not clear how useful such a type would be in practice.

Actual implementation work should be very easy. Core and sync both provide the necessary primitives and expose them as APIs already. We just need to figure out what this feature should look like as a public API, which will affect what support object store provides for bindings (if any).

@jpsim @bdash @tgoyne

**Comment by astigsen** [comment]
The reasoning for making it the same type is that you also want to be able to reset the counter to specific state, which is functionally the same as updating an integer value.

**Comment by jpsim** [comment]
I don't see introducing a new type as a clear cut solution.

### New Type Advantages

* No way to accidentally call "set", overriding past history of increment/decrement
* Communicates intent of how a property should be used at the type system level
* Add/Subtract can be methods directly on the type rather than on the parent object

### Int Type Advantages

* No need to add a new type and protocol (e.g. `RLMCounter`)
* Works with existing code that expects integers without "unwrapping"
* It's useful to be able to "reset" the counter. There's no reason to force that reset value to be zero, so if it was on a new type, that'd just be `.reset(_ value: Int)`. At that point, it's functionally equivalent to just `prop = value`.

---

Based on this, I think I prefer the new type option, but only slightly. I may be missing arguments on either side here too. In any case, we should sketch out what we want the final public API would look like.

**Comment by austinzheng** [comment]
Any thoughts, @cmelchior @kristiandupont @AndyDentFree?

**Comment by cmelchior** [comment]
Java doesn't have an option. It will have to be a new type since we cannot add functionality to existing classes. Something like 

```
public class RealmCounter implements Comparable {

  // Constructors
  public RealmCounter(String val);
  public RealmCounter(long val);
  // Possible others?

  // Only required operation, arguable if substract is needed or not
  public void add(long val);

  // Convert to java primitive types. Integer in Java also have to float/double conversion
  public int intValue();
  public long longValue();
  public short shortValue();
  public float floatValue();
  public double doubleValue();

  // Comparable
  public int compareTo(RealmCounter other);
}
```

You can probably argue how many utility methods to add to it, but since this will be a class we can probably ship a MVP of it and add more methods later depending on feedback. I assume we are bounded by 64 bit values and don't support e.g. BigInteger. That is probably of questionable value anyway.

Some inspiration:
From Riak: https://github.com/dclements/riak-java-crdt/blob/master/src/main/java/com/readytalk/crdt/counters/CRDTCounter.java

This one has different classes for various semantics (seem a bit overkill to me).
https://github.com/dclements/riak-java-crdt/tree/master/src/main/java/com/readytalk/crdt/counters

**Comment by cmelchior** [comment]
We also have to consider how migrations on this work, e.g it should be possible to change the field from a counter field back to a normal int. That seems hard to do if you overload `Int`?

**Comment by jpsim** [comment]
> We also have to consider how migrations on this work, e.g it should be possible to change the field from a counter field back to a normal int. That seems hard to do if you overload Int?

Since counter-style add/subtract are instructions on core int columns, there's no impact on schema between a property of type `Int` or "Counter". Even if we go with a custom Counter type, we can not require a schema bump when converting a property from Int to Counter.

> Java doesn't have an option. It will have to be a new type since we cannot add functionality to existing classes. Something like

We also can't realistically have these methods on `Int` in cocoa, which is why I was referring to the "existing type" option above as having its add/subtract methods on the parent object, which is ugly as sin:

```swift
extension Object {
  func add(amount: Int, toProperty: String)
  func subtract(amount: Int, fromProperty: String)
}
```

**Comment by cmelchior** [comment]
> Since counter-style add/subtract are instructions on core int columns, there's no impact on schema between a property of type Int or "Counter". Even if we go with a custom Counter type, we can not require a schema bump when converting a property from Int to Counter.

I think we need to find a way to do that then, as this is a semantic breaking change with very surprising consequences. Consider a case for sync where 

```
public class Person extends RealmObject {
  // V1
  public int muffinsEaten;

  // V2
  public RealmCounter muffinsEaten;
}

muffinsEaten = 4 // V1 device  
muffinsEaten.add(4) // V2 device -> state is now 8
muffinsEaten = 6 // V1 device
muffinsEaten.add(2) // V2 device -> state is now 8
```

In my mind, this is a breaking schema change and as such shouldn't be allowed for Sync and would require a schema bump for non-synced Realms.

**Comment by AndyDentFree** [comment]
In .NET we have no problems using a system structure type [DateTimeOffset](https://msdn.microsoft.com/en-us/library/system.datetimeoffset%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396) and so I see no problems with a `RealmCounter` which provides operator overloading to inc and dec.

**Comment by austinzheng** [comment]
I think we have a consensus, then, that it should be a separate type? If we want users to be able to arbitrarily set a value (bad idea) we can always expose a `reset()` API that is functionally identical to `set()`, but at least provides the intended semantics.

**Comment by AndyDentFree** [comment]
Separate type, backed by int property so no different storage needed, works for me.

ObjectBrowser needs to be aware of the type as we may want to _warn_ people if they try to arbitrarily set it (not **ban** arbitrary setting because that would enrage devs).

**Comment by cmelchior** [comment]
> ObjectBrowser needs to be aware of the type as we may want to warn people if they try to arbitrarily set it (not ban arbitrary setting because that would enrage devs).

Not sure what you are talking about here?

**Comment by AndyDentFree** [comment]
Even though backed by an int, ObjectBrowser needs to have rules aware this is a counter so it can warn people when they go to reset.

**Comment by jpsim** [comment]
No, I don't think tracking in metadata table that the "int" column is really a "counter" is helpful.

**Comment by AndyDentFree** [comment]
>No, I don't think tracking in metadata table that the "int" column is really a "counter" is helpful.

I have gone back over the entire discussion. It seems we are as a consensus leaning towards a separate type for bindings.

I'm not sure if Counter behaviour needs to be different from int for sync also, as the comments suggest it will map to something like `counter_inc`, `counter_dec` and `counter_set` instructions which could as well be applied to integer columns.

If it is a separate type for bindings then of course they need to be able to infer that from the schema to support dynamic actions.

However, if the underlying storage is just an int, I agree with @jpsim that it doesn't seem to require a schema bump to move between Counter and int types - it's behavioural change rather than data change. However, I do think it's dangerous to start adding special case logic or _some_ property type changes so that inclines me back to treating it like any other schema change - a version bump is required.

I have spent so much of my career being a toolsmith that I always tend to think of people writing tools against the engine. Not just for our ObjectBrowser but, increasingly, as we move into the Enterprise Space we will have people wanting to write their own tools - this has been a driver behind at least two of the Win32 desktop requests.

**Comment by simonask** [comment]
FYI, Sync does not treat counters and integers as separate types. You can perform an `AddInteger` instruction on any integer column, and `SetInteger` just resets the counter. These instructions can be freely mixed with no odd effects (`SetInteger` is equivalent to resetting the counter, as others have mentioned).

**Comment by kristiandupont** [comment]
In that case, at least in .net, I'd say it would be cleanest to simply add dedicated `Add` and `Set`functions that works on integer columns.

**Comment by cmelchior** [comment]
I strongly disagree. Counters should not be treated as simple integers in the bindings. Just because Core can optimize the underlying representation doesn't mean that has to surface to the bindings (Just like the old Date implementation wasn't an Integer either). The difference in semantics requires treating it as a separate type IMO with the impact it has on API/Browser/Etc.

**Comment by jpsim** [comment]
I think we're all in agreement to expose a Counter type in the bindings. What's less clear is whether or not this should be annotated in the Realm's metadata, and if so how? Introducing a new way to specialize or refine types in a schema would need to be done with some consideration of how it should scale in the future too.

**Comment by austinzheng** [comment]
@simonask Is it intentional that there's no `RowFuncs` implementation of `add_int()`? It's a bit cumbersome to get the table from a column from the object store to use the instance method on `Table`.

**Comment by cmelchior** [comment]
One way could be to have a new mapping table between types understood by the OS and Core types:

```
// Enum defined in the Object Store, enumerating all specialized types
enum RealmType {
  Counter, IncrementOnlyCounter, ....
}

// New meta data table
class | field | type
-----
Party | guests | Counter
````

Being able to annotate Table Columns would be nicer, and could be used for other things as well, but that would require some help from Core I would assume (and probably a file-format upgrade as well).

**Comment by simonask** [comment]
Please remember to consider that the Realm's metadata is also being synchronized. If you want to treat counters truly as separate types, the migration to upgrade existing integer columns will be schema-breaking, whereby it has to be disallowed for sync'ed Realms for the time being.

This would be a case where that limitation is entirely artificial and solely motivated by our desire to prevent users from shooting themselves in the foot. I'm questioning how this benefits the user, exactly. If the user does it wrong, there wouldn't be a way for them to fix their code later on (because it would require schema-breaking migrations).

In short, I don't really think there is a good reason to persist the counternance ( 😄 ) of an integer field in the schema. Just let it be up to the app developer to use the binding's `Counter` type if they wish to have a counter.

**Comment by cmelchior** [comment]
>  If the user does it wrong, there wouldn't be a way for them to fix their code later on.

So far we have been going to great lengths trying to prevent users from shooting themselves in the foot, and I don't see how preventing switching between Integer and Counter are any different than not allowing people to migrate a column from Integer to Strings or a host of other schema breaking migrations. Sync schemes are just very restrictive at the moment (See e.g. https://github.com/realm/realm-sync/issues/400)

 As my example https://github.com/realm/realm-object-store/issues/357#issuecomment-275613063 showed. The semantics for mixing Integer and Counter across app versions will be completely unintuitive which strongly suggests you shouldn't be able to switch types for synced Realms.

**Comment by simonask** [comment]
I would say that unintuitive results is still better than exclusion from synchronization altogether. It's easy to explain what happens, and to instruct users to just use a different type for the field in their model. It's not so easy to guide them through migrating the schema with destructive changes. :-)

**Comment by cmelchior** [comment]
On that we disagree then. It might be "easy" to explain, but my default stance is that we should not introduce bugs in users app if they don't read the docs (which most of them won't). 

Also, we are not preventing them from synchronizing. They need to create a new field with the Counter type (which they must do for other type migrations as well), which will also make the behavior difference explicit in their code base.

**Comment by simonask** [comment]
> Also, we are not preventing them from synchronizing. They need to create a new field with the Counter type (which they must do for other type migrations as well)

This is just as bad, and should be avoided when possible. In all other instances where we ask users to do this, it is because there is no good alternative. Here it would be completely artificial.

> which will also make the behavior difference explicit in their code base.

That will be the case no matter what (for bindings to languages that don't have operator overload - note that a C++ binding in the future probably doesn't need or want a separate `Counter` type).

**Comment by cmelchior** [comment]
> This is just as bad, and should be avoided when possible

Yes it is bad, but somehow I feel you don't fully acknowledge my argument that semantic breaking behavior is just as bad as API breaking behavior. If we don't agree on that, we probably won't agree on how to approach Counters.

**Comment by simonask** [comment]
Yes, I don't agree with you on that. :)

I think it is much more important to provide users with a way forward. Temporarily wonky behavior due to initial misunderstanding is better than being permanently locked to a data model that is less than ideal. I really think breaking the user's schema should be a very last resort.

**Comment by cmelchior** [comment]
Fair enough :)

I would much prefer https://github.com/realm/realm-sync/issues/400 as a way to handle breaking schema changes though since it would enable people much more flexibility with regards to how they approach schema changes.

**Comment by simonask** [comment]
Yes, we definitely need a better long-term solution for breaking schema-changes. :+1:

**Comment by AndyDentFree** [comment]
I was just thinking about this and wondering where the original  user scenarios come from.

In the process, I had a couple of ideas which may help the debate:

1. if we want a Counter type to function as a robust _counter_ which you increment/decrement and occasionally reset then we want to **make _accidental reset_ hard.** This led the conclusion that we should _disallow assignment from an integer,_ which makes it more obviously a separate type.
1. Having _merge-safe arithmetic_ with add/del as operations rather than just assigning their result, seems generally useful. It then seems weird we don't therefore allow it on floats and that we lack multiplication.

I was thinking about one possible scenario of financial balances (I've worked on a lot of accounting software). Traditionally, a balance is calculated as a result of summed journal entries creating a delta from a given starting point. 

Our Add/Del operations could be seen as _unannotated jounal entries_ as they are just the operation without accompanying explanation. However, they can also be an optimisation - say we also log immutable journal entries but in the same Realm transaction have these math operations. The net effect, regardless of order, should be the same without the app having to sum its accounts.

**Comment by austinzheng** [comment]
Thinking about original user scenarios are a good idea. My inclination is that #1 is more useful. However, maybe the sample apps folks might have more ideas.

**Comment by astigsen** [comment]
I would like for us to really consider if this needs to introduce a new type. In general we will want to continue making it possible for users to express the intent of their changes in more and more detail.

As an example of this you can see strings. Today we only (in the bindings) allow strings to be set, but in the future we will also allow inserts and deletes within strings, and later possibly moves and swaps. For all of these new operations there should be no need for new types as they are perfectly backwards compatible (apps that only support full string changes will only see the end result of these sub-string modification, so from their perspective they are equivalent of an atomic update).

Today with integers you can only set them atomically, but allowing people to express their intent more clearly seems valuable. If you want to set it you should do that, but if what you are really doing is to increment or decrement the the integer, you should be able to express that in a way that records the intent. Just as with strings it ought to be totally backwards compatible. An older app should not have to be aware of how the integer was modified, it will just see the end result.

If we start adding new types every time we give the user more fine-grained options for expressing how they use them, we could easily end in a combinatorial explosion of types.

**Comment by cmelchior** [comment]
As I noted in https://github.com/realm/realm-object-store/issues/357#issuecomment-275349370, Java doesn't have any option in this regard, we will need to introduce a new type since we cannot add functionality to existing classes due to language constraints.

I think we all agree that allowing people to express their intent more clear is better (hence adding Counter in the first place), I'm just extremely concerned that allowing you to mix `set` with `add/increment` semantics will not behave in an intuitive way (for the end user), which is the entire reason I'm advocating for making them different binding types that require migrations.

The simple case
```
v1 app (int)       v2 app (counter)
0                  0
set(2) = 2
                   2
                   add(1) = 3
                   add(1) = 4
4                   
----------- disconnect -----------
                   add(1) = 5
                   add(1) = 6
set(2) = 2
----------- connect -------------- 
2                 2
// old client will make it look like new client is loosing data
```

The above is clearly backward compatible and will converge given the rules for our merge algorithm, but any old client will just override the intent expressed by the Counter. 

I'm pretty sure that isn't what the user wants if he is switching to a Counter in the first place. And yes, we could just document the behavior like Simon is advocating for but if we strive for the Principle of Least Astonishment then that is the wrong approach IMO. That intent is better expressed in the type system with `Counter` and `Int` as separate types. That way developers will need to handle the different semantics _before_ the app is shipped to customers.

**Comment by simonask** [comment]
Sorry, maybe my head is buried in OT, but that result does not seem unintuitive to me at all. Explaining it is as easy as "the problem you have is that old clients reset the counter instead of adding/subtracting". I still don't see how it's helpful to break compatibility to prevent this situation.

I agree, by the way, that the binding should expose counters as a separate type in the data model. I don't think there's a good enough reason to make `Counter` and `Int` incompatible.

**Comment by cmelchior** [comment]
Yes, explaining the problem is easy. But how would you suggest the user going about fixing it? Since they cannot prevent old clients from syncing to new ones currently?

**Comment by simonask** [comment]
Bugs in old apps that get synchronized to newer versions is a much broader general issue that I don't think we can solve for people, at least not right now. Schema versioning with distributed migrations may be the right answer, but that definitely needs more thought before we can start designing on the assumption that it exists, IMO.

You could also argue that the amount of wonkiness introduced to the user experience by forcing developers to add a new `Counter` field in their new app version is pretty much equivalent to the amount of wonkiness introduced by old apps updating a counter field. The latter is just less work for us. ;-)

**Comment by astigsen** [comment]
I still think it is a slippery slope. What happens when we start getting requests for setting and clearing individual bits on integers (which also commute)? We already have people asking for support for querying against bit patterns (which is a standard NSPredicate feature). Do we then have to add another type?

**Comment by jpsim** [comment]
Let's all remember that language constraints sometimes prevent us from doing "the obvious" thing.

For example, you can't override the increment/decrement operators in ObjC or Swift, which means that a user could do `obj.intProp++` and expect that to be an increment operation where in reality it'd be a (re)set operation. A new type prevents that from happening if the integer unwrap is a get-only operation.

**Comment by nirinchev** [comment]
I think @astigsen referred to exposing the counterness in the schema, not the binding. I think we all agree that none of the bindings can expose this functionality using system types.

**Comment by AndyDentFree** [comment]
if we don't have it in the schema, how does a binding know to use the type? I am thinking of enterprise settings where people write multiple language apps and if you have compatible types in the schema with no indication of a different semantic, they can build an app with an `int` access to a `counter`. Also, it impacts dynamic opening and browser behaviour.

**Comment by astigsen** [comment]
> they can build an app with an int access to a counter

What would that mean? A `counter` _is_ just an `int`. The only difference is which operations you choose to do on it.

> Also, it impacts dynamic opening and browser behavior.

There is no change in behavior when reading the values. So for browsing the data there should be no change. For editing the browser only supports setting values now, so when more ways of changing them becomes possible, we could consider making them available in the browser as well (that goes for strings a well).

One reason I like the idea of just augmenting the existing type with more operations, is that it is not always easy to predict ahead of time what you want to do with them. You might initially only need to set full values, but at some time later you might see a need to do increments and decrements. It would be nice for that to just work, without having to involve a migration that breaks old apps.

The point here is that the old apps explicitly communicate the intent that they are _setting_ values. So as long as that was the correct intention at that point, it should still be valid and the app should keep working just fine. If we force a migration that changes the type of an existing property, just because they found out that there are now places where an increment operation is needed, then we break all the old versions of the app.

**Comment by AndyDentFree** [comment]
If we expose a custom type in the bindings, then operator overloading such as in [C#](https://msdn.microsoft.com/en-us/library/sa7629ew.aspx) and [Swift](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/AdvancedOperators.html) can capture operators that are performing an increment or decrement. We could have functions in Java such as `Add` or `Inc`.

If the property is exposed only as an integer then **any user operation on it is a reset**.

**Comment by cmelchior** [comment]
@astigsen Think of it the other way. Is there _any_ scenario where it would be the correct behavior if the field was an Integer in V1 of the app and a Counter in V2 of the app? (especially given the example in https://github.com/realm/realm-object-store/issues/357#issuecomment-278286497). I cannot come up with any, but if you have a use case I would be all ears.

If not, that is a strong argument for keeping them as separate types. There is also a reason that ideas like Hungarian notation came into existence and languages like ML allow you to define custom datatypes. Types carry information about intent far better than implicit knowledge.

E.g I cannot tell from an `Integer` if it is a Date, Bit pattern, Counter, Color or what it represents. In Android they added a huge number of annotations precisely because `Integer` is so unprecise: https://developer.android.com/reference/android/support/annotation/package-summary.html 

I totally get the concern about type explosion, but IMO having e.g. just an`RealmInteger` type with `set/increment/decrement/setBit` methods would only benefit us because it would be less code to write, but users would suffer from unclear code and impose a hidden requirement that they cannot mix semantics across app versions anyway.

Note: My concern is only for synced Realms. Non-synced Realms would work fine with just one type, but I'm pretty sure we don't introduce different rules for sync/non-synced Realms.

**Comment by nirinchev** [comment]
To me breaking the schema just in case feels like overprotecting. As a developer, it would be clear to me that when old versions set an integer they'll reset my counter. If we make the new counter type implicitly convertible, they'll still be able to assign to an int in newer versions, so that kind of protection seems insufficient/unnecessary.

**Comment by cmelchior** [comment]
A counter set/reset is different than a pure integer set. A counter (re)set would be equivalent to `int.add(target - int.get())`, and we would probably expose that on a counter as an utility method anyway. It would merge very differently than a pure integer set.

**Comment by simonask** [comment]
> A counter (re)set would be equivalent to int.add(target - int.get())

That's not true, and that would produce very unexpected results. A counter reset is exactly equivalent to assigning to the integer (and will be treated as such by the merge rules already).

**Comment by cmelchior** [comment]
Yeah, your right, I was imprecise. There is a difference between `set` and `reset`.

**Comment by simonask** [comment]
No, there is no difference between set and reset. :)

**Comment by cmelchior** [comment]
Arguing semantics, but I disagree. A `reset` sounds like a destructive operation (which it is), `set` is an operation that is fundamentally at odds with how a counter is supposed to work. But that is probably not relevant for the discussion about types 😄

**Comment by simonask** [comment]
Isn't it exactly at the heart of the discussion, though? I really don't see a reason to distinguish between the two things.

**Comment by cmelchior** [comment]
Please provide one single use case where it would make sense to mix `increment/set` between app versions? Because if such a case doesn't exist, I don't see why we should allow it?

**Comment by nirinchev** [comment]
Any sort of imprecise counter where loss of data is acceptable would be perfectly fine example. E.g. total number of steps - the rare case where an older version (implementing a poor man's counter by doing `steps++`) is offline, assigns new values, then comes back online and overwrites increments from a newer version will lead to a data loss, but in my opinion would be acceptable.

There are many other approximate vanity counters that would be fine to reset - e.g. number of games played, total amount of gold in a clicking game, etc. Resetting those will not corrupt state (e.g. because the user will have already consumed the gold). As a developer, I could see myself migrating to the new counter type and just ignoring the inconsistencies that might happen due to older versions syncing after being offline for a prolonged period of time.

Now, for the counter-question, can you provide a legitimate recipe for handling breaking the schema from a developer's point of view? 😛

**Comment by astigsen** [comment]
> Please provide one single use case where it would make sense to mix increment/set between app versions?

There are plenty. Obvious ones are everything related to inventory control. You can easily have the first version of an app used for counting inventory, where they manually count the number of products in each category and enter the number in the app. Then later they find out that they would like the number to also be somewhat up-to-date between the full recounts, and it really sucks for each employee to have to recount the entire category every time they add or remove an item, so they either update the app of make a new app for employees where they can increment or decrement the count when adding or removing product. They will still occasionally do a full inventory where they count everything and enter the verified values.

Another one I have seen many times is participant tracking at events, where you have people doing manual counts at each entryway (usually done today with a manual ticker) to get an approximation of how many people are at the venue or in the bus, and then when the doors are closed they do a real count to update it to the precise number (my mom is a tourist guide, so I have seen this process happen thousands of times :-)).

But I think it is more relevant to look at it the other way around. How often do you do a `++` or `+=` on integers in regular programming? What are the examples of this being done, where if it happens in concurrently, you would not want to to merge as increments (as opposed to the end results just overwriting each other). To me it is just about exposing a set of semantics that you do anyways on integers, but which needs to be recorded to make sense in a concurrent setting.

**Comment by cmelchior** [comment]
@nirinchev Fair enough. I did not consider fuzzy counters although they seem more like acceptable data loss than something you would desire. 

Regarding migrations, you would do what you need to do in any other breaking change: Create a new field and copy the value over. Right now our migration story is pretty weak though as we do not run migration blocks on synced Realms, so people would have to track it manually, and yes that would mean data loss when doing the migration as well, but at least you would know it happened instead of it coming as a surprise (at which point it is too late to fix).

In any case, no case in arguing this to death, so lets just start with a `RealmInteger` class or similar like @jpsim suggested. It can then encode all the custom methods we want on it and be fully compatible with a standard `int`. Then we just have to be exceedingly clear in the docs how it actually works.

At least it will keep the number of classes low and would also have nice symmetry with e.g `RealmString` for substring manipulation. If enough people get burned by it, I guess we can re-evaluate.

**Comment by austinzheng** [comment]
@jpsim So, to sum up, at the object store levels `Integer` properties should get access to the add functionality. Then bindings will implement a new Counter type backed by an Integer, just like the regular Integers. Is that correct?

**Comment by jpsim** [comment]
Yes, that's optimal in my opinion.

**Comment by cmelchior** [comment]
I disagree. If we want to have a super-int. It shouldn't be named `RLMCounter`, especially if it will also gain bitwise operations as hinted by @astigsen. If it is just intended as "Int with special capabilities", `RLMInteger` or similar would be preferable. It would also make it more apparent that you can switch between `Integer` and `RLMInteger` at will.

**Comment by austinzheng** [comment]
@cmelchior You are referring to how bindings choose to expose this functionality to end users, right?

**Comment by jpsim** [comment]
> I disagree. If we want to have a super-int

A counter is exactly a super-int, though. Add/Subtract/Reset. It'd be incredibly confusing to users to call that an `RLMInteger` or similar when `int` is also an option.

**Comment by cmelchior** [comment]
Yes, but it bleeds down to OS as well, e.g. it seems weird to add a `Counter` and `BitPattern` (or whatever name) class in OS if you combine them in the bindings.

**Comment by austinzheng** [comment]
I wasn't planning on adding additional object store types, just exposing the `add_int()` Core API to bindings through the object accessor mechanism.

**Comment by cmelchior** [comment]
@jpsim I would agree if we only had Counter functionality, but how I read https://github.com/realm/realm-object-store/issues/357#issuecomment-278411081 is that will most likely change, so it can have multiple special methods not just related to counters.

**Comment by austinzheng** [comment]
I'm beginning to put together a PR: https://github.com/realm/realm-object-store/pull/375. I want to test it (especially since templates aren't instantiated unless you actually use them somewhere...), but it doesn't seem like we exercise any of our existing object accessor functionality in our object store tests. So I have to think about it some more.

**Comment by bmeike** [comment]
The core (as in "bulk", "central") implementation of RealmIntegers is now in Java mainline ( https://github.com/realm/realm-java/pull/4953 )

**Comment by bmeike** [comment]
Counters are now implemented in the Java bindings: #4266

**Comment by bdash** [comment]
@austinzheng, did we determine whether anything needs to be done in the object store for this? Java and .NET seem to have been able to support counters without any object store support. Is that because they're not using object store's accessor logic?

**Comment by austinzheng** [comment]
Object store support is unnecessary for counters for Cocoa either. I originally made this issue in the object store repo because I thought we might need underlying changes.
